### ISSUE TITLE
Locking Layers

### ISSUE THREAD

**Issue by MosGeo** [napari/napari#3466]
## 🚀 Feature
This is my initial thinking on this feature. I would love feedback on the best approach. I propose the ability to lock and unlock layers. This would include the ability to control it programmatically and through GUI.

## Motivation
My main motivation for this is to prevent user from closing the layer while live imaging capturing is happening. Another is locking layers while some image processing is happening.

## Pitch
An icon beside each layer (besides the eye visibility icon) can be inserted to lock the layer. There are two types of locks that can be used:
1. Layer where locking can be controlled by the user. If a layer is locked by the user. No plugin can change the data.
2. Layer where locking cannot be controlled by the user (including optional deletion prevention). Only code can unlock/delete this layer. These are useful for some plugins like live webcam plugin. For these layers, a helper function can set the tooltip to highlight what plugin/process is locking the layer.

## Alternatives
Currently to lock layer, you can modify the editable property. This prevent some functionality but you can still delete the layer. In the GUI, there is no way to lock layers.

## Additional context
As my main functionality that I am considering is live preview, I go back and forth on the functionality. For example, maybe i should just stop the camera once the layer is deleted. That is possible right now (and this is what I have implemented). Still, I think having locking feature would be useful for other scenarios.

**Comment by MosGeo** [comment]
@haesleinhuepf Would love to hear your opinion on this? My current implementation for live preview prevent editing, hides all other layers, and recenter the viewport based on the camera resolution.

Currently, it doesn't change back the visibility of the layers to their original configurations. Also, there is no guarantee that the layers even exist so it would need further checks (which is easy to implement)

**Comment by sofroniewn** [comment]
There has been some good discussion of this in https://github.com/napari/napari/issues/1570, I think we'd be more than happy to have someone pick up this work and implement it. I'm happy to help guide this a bit too

**Comment by MosGeo** [comment]
Here is my initial thinking on the subject after reading #1570

# Concept
Design a general locker to hold keys (dictionary of keys) for any object. The locker would be attached to the object (mainly layer at the start but it is generic). It is basically the dictionary concept in #1570.

# Details
- Each key holds its status (locked or unlocked), a value (e.g., the maximum number of layers), and a mode (is it locked from the user perspective only or code or both. I don't know if the mode is needed or just agree to lock things from the user perspective (interface)  only and code can do anything?
- The locker is non-binding, i.e., it is up to the developer to query it and check things before modifying the properties. 

# Possible implementation 

```python
"""
A generic lock class to be used as a locking mechanism. This lock is a non-binding agreement, i.e.,
it is up to the developer to use it and make sure the locks are honoured.
"""
from typing import Dict, Optional, Any
from enum import Enum

class LockMode(Enum):
    """
    The lock modes can be one of three value
    1. Soft: Closer to suggestion that should be followed by the interface developer.
    2. Hard: both code and interface cannot change the value
    """
    SOFT=0
    HARD=1

class Lock():
    """
    A lock consists of a three components:
    1. A boolean to house the status of the lock (open, locked)
    2. A value of the lock, for example, an attribute maximum can be locked to certain number
    3. A lock mode to desginate the type of lock (soft (user), or hard (code and user))
    """
    is_locked:bool
    value:Any
    lock_mode:LockMode
    comments:str

    def __init__(self,
                 value:Any=None,
                 lock_mode: LockMode=LockMode.HARD,
                 is_locked:bool=True,
                 comments:str=""):
        """Initializes a lock"""
        self.value = value
        self.lock_mode = lock_mode
        self.is_locked = is_locked
        self.comments = comments

    def __str__(self):
        return f"Lock value: {self.value}, lock mode: {self.lock_mode}, locked: {self.is_locked}"


class Locker():
    """
    A generic class to be used as a locking mechanism. It contains a collection of locks
    for different attributes.
    """
    _lock_dictionary:Dict[str, Lock]

    def __init__(self, lock_dictionary:Optional[Dict[str, Lock]]=None):
        """Initializes the lock"""
        if lock_dictionary is not None:
            self._lock_dictionary = lock_dictionary
        else:
            self._lock_dictionary = {}

    def add_lock(self, attribute:str, 
                 value:Any=None,
                 lock_mode:LockMode=LockMode.HARD,
                 is_locked:bool=True,
                 comments:str=""):
        self._lock_dictionary[attribute] = Lock(value, lock_mode, is_locked, comments)


    def is_locked(self, attribute:str, lock_mode:Optional[LockMode]=LockMode.HARD)->bool:
        """Check if there is a lock on a specific attribute"""
        if not self.has_attribute(attribute):
            return False
        elif self._lock_dictionary[attribute].lock_mode == LockMode.HARD:
            return self._lock_dictionary[attribute].is_locked
        elif lock_mode == self._lock_dictionary[attribute].lock_mode:
            return self._lock_dictionary[attribute].is_locked
        else:
            return False


    def has_attribute(self, attribute:str):
        """Checks if the lock has a specific attribute"""
        return attribute in self._lock_dictionary


    def lock(self, attribute:str):
        """Locks a specifc attribute"""
        self._get_lock(attribute).is_locked = True
            

    def unlock(self, attribute:str):
        """Unlocks a specific attribute"""
        self._get_lock(attribute).is_locked = False

    def list_locks(self):
        """Returns the list of locks available"""
        return list(self._lock_dictionary.keys())


    def _get_lock(self, attribute:str)->Lock:
        """Returns the lock value"""
        return self._lock_dictionary[attribute]
```

# Impact
Because it is a dictionary, the implementation of locking things can be gradual. For example, you can implement the locking for certain property only initially and after a couple a months, you can implement another one. 

Would love to hear what you think.

**Comment by sofroniewn** [comment]
Thanks for the detailed proposal here.

> The locker is non-binding, i.e., it is up to the developer to query it and check things before modifying the properties.

I guess this is something that we can make sure we respect internally, but we might not find high compliance from plugins/ external API calls.

Another approach might be to add something to our EventedModels, maybe a LockableEventedModel which would check the "lock" on setting and actually prevent setting if the lock was applied. We still need to make our layers evented models, but it is a goal. Maybe in pydantic land, these locks are a little like making attributes "read-only" or "frozen". I know you can do that on construction, I'm not sure if you can change "frozen" status dynamically? cc @tlambert03 

Maybe some combination of @tlambert03 @jni @kevinyamauchi @alisterburt or @andy-sweet might be able to give more feedback on this.

**Comment by tlambert03** [comment]
> The locker is non-binding, i.e., it is up to the developer to query it and check things before modifying the properties.

yeah, this feels like a bad idea to me.  i suspect that approximately no one will check this, making it effectively useless.  A lock should be a lock.

I think this needs to be implemented in `__setattr__`.  As @sofroniewn mentioned, pydantic does provide a frozen/allows-mutation switch, however this is set at the class level (not the instance level), which means we'd still need to do some mucking about to use pydantic's immutability.  so we're probably better off implementing our own API in our `setattr` (not to mention the fact that our layers aren't yet pydantic models, so that's all kind of moot anyway).

Anyway, for me, probably the primary design criterion here is that locking isn't just suggestion for API users (plugins, etc...) but rather a hard rule.  (How frustrating would it be for an end user to lock something, just to have a plugin override it because it didn't "check the lock")

**Comment by MosGeo** [comment]
@sofroniewn @tlambert03 Thanks for the guidance. 

1. Using __setattr__ is possible to check for locks simply using the below for the base layer. I've tested it and I was able to lock the name property for example. if it is a lock, we can raise an error or a warning or do something.

```python
    def __setattr__(self, __name: str, __value: Any) -> None:
        if  self.locker.is_locked(__name, lock_mode=LockMode.HARD):
            print(f"{__name} is locked")
        else:
            return super().__setattr__(__name, __value)
```

2. I am still not sure what is the best way to have two lock modes: actual lock, and a lock to stop only interface related changes. I believe having a soft lock is important to give fixability to plugin developers (e.g,. my use case outlined above of live viewer). Right now, i changed it to hard and soft lock modes. What are your opinion on this?

The workaround of having hard code and soft interface lock would be the need to unlock, perform operation, and relock whenever you want to make a change.

I can put a draft fork for this to test out if needed at this stage.

p.s. I updated the possible implementation slightly.

**Comment by sofroniewn** [comment]
> I am still not sure what is the best way to have two lock modes: actual lock, and a lock to stop only interface related changes. I believe having a soft lock is important to give fixability to plugin developers (e.g,. my use case outlined above of live viewer). Right now, i changed it to hard and soft lock modes. What are your opinion on this?
>
> The workaround of having hard code and soft interface lock would be the need to unlock, perform operation, and relock whenever you want to make a change.

It does make sense to me that a plugin dev might want to lock a property from being editable in the GUI, but keep letting it be edited by the plugin itself. I'll have to think more about the different possible implementations, but I think I get the use case

**Comment by Czaki** [comment]
One option to workaround this is to keep a local reference to the layer (in the widget) and add it again if it is missed in layertlist. Current napari release allow adding again removed layer.

**Comment by MosGeo** [comment]
> One option to workaround this is to keep a local reference to the layer (in the widget) and add it again if it is missed in layertlist. Current napari release allow adding again removed layer.

Yes. That is one option for this specific use case. As napari is looking for a general solution for any property (per #1570), I thought it would be good to design the more general solution that would encompass this use case.

**Comment by Czaki** [comment]
To implement properly permissions there is a need to implement also ownership to release permission when controlling widget/object was deleted, or just some sudo mode. To allow user delete a layer, without the need of napari restart, if there was some error in plugin code which corrupt object.

**Comment by MosGeo** [comment]
Ok, below is the latest prototype that I have. Three items still are still in the investigation phase:

1. It is now a hard lock by default as suggested by @tlambert03 and @sofroniewn.
2. I added lock mode for locking values too. For example, value can be locked to a specific choices or a range.
3. As @Czaki  excellently pointed out, there should be some kind of override (either manual or automatic) in case of bugs. Nothing is implemented yet in the code below but I am thinking something that holds the source and if the source is `None`, we can assume that the plugin crashed and unlock things? Does the plugin widget even goes `None` when closed? I remember I read a pull request or an issue about closing plugins somewhere. One thing that I am trying to avoid in this solution is injecting any QT into this class (separation of GUI and Core).

```python
"""
filename: napari.utils.lock.py
A generic lock class to be used as a locking mechanism. This lock is a non-binding agreement, i.e.,
it is up to the developer to use it and make sure the locks are honoured.
"""
from typing import Dict, Optional, Any
from enum import Enum

class LockMode(Enum):
    EXACT=0
    IN_LIST=1
    IN_RANGE=2
    LARGER_THAN=3
    SMALLER_THAN=4

class Lock():
    """
    A lock consists of components:
    1. A boolean to house the status of the lock (open, locked)
    2. A value of the lock, for example, an attribute maximum can be locked to certain number
    3. A lock mode to desginate the type of lock (soft (user), or hard (code and user))
    4. comments
    """
    is_locked:bool
    value:Any
    is_hard_lock:bool
    comments:str

    def __init__(self,
                 value:Any=None,
                 is_hard_lock:bool=True,
                 value_lock_mode:LockMode=LockMode.EXACT,
                 is_locked:bool=True,
                 comments:str=""):
        """Initializes a lock"""
        self.value = value
        self.is_hard_lock = is_hard_lock
        self.value_lock_mode = value_lock_mode
        self.is_locked = is_locked
        self.comments = comments

    def __str__(self):
        return f"Lock value: {self.value}, lock mode: {self.lock_mode}, locked: {self.is_locked}"


class Locker():
    """
    A generic class to be used as a locking mechanism. It contains a collection of locks
    for different attributes.
    """
    _lock_dictionary:Dict[str, Lock]

    def __init__(self, lock_dictionary:Optional[Dict[str, Lock]]=None):
        """Initializes the lock"""
        if lock_dictionary is not None:
            self._lock_dictionary = lock_dictionary
        else:
            self._lock_dictionary = {}

    def add_lock(self,
                 attribute:str, 
                 value:Any=None,
                 is_hard_lock:bool=True,
                 is_locked:bool=True,
                 value_lock_mode:LockMode=LockMode.EXACT,
                 comments:str=""):
        self._lock_dictionary[attribute] = Lock(value, is_hard_lock, value_lock_mode, is_locked, comments)


    def is_valid_value(self, lock:Lock, value:Any):
        if lock.value_lock_mode == LockMode.EXACT:
            return False
        elif lock.value_lock_mode == LockMode.IN_LIST:
            return value in lock.value
        elif lock.value_lock_mode == LockMode.IN_RANGE:
            return value>= lock.value[0] and value <= lock.value[1]
        elif lock.value_lock_mode == LockMode.LARGER_THAN:
            return value>=lock.value
        elif lock.value_lock_mode == LockMode.SMALLER_THAN:
            return value<=lock.value
        else:
            return True


    def is_locked(self,
                  attribute:str,
                  is_hard_lock:bool=True,
                  value:Any=None)->bool:
        """Check if there is a lock on a specific attribute"""
        
        if not self.has_attribute(attribute):
            return False
        
        lock = self._lock_dictionary[attribute]
        if not lock.is_locked:
                return False
        
        if lock.is_hard_lock is True or is_hard_lock is False:
            if self.is_valid_value(lock, value):
                return False
            else:
                return True
        return False


    def has_attribute(self, attribute:str):
        """Checks if the lock has a specific attribute"""
        return attribute in self._lock_dictionary


    def lock(self, attribute:str):
        """Locks a specifc attribute"""
        self._get_lock(attribute).is_locked = True
            

    def unlock(self, attribute:str):
        """Unlocks a specific attribute"""
        self._get_lock(attribute).is_locked = False


    def list_locks(self):
        """Returns the list of locks available"""
        return list(self._lock_dictionary.keys())


    def _get_lock(self, attribute:str)->Lock:
        """Returns the lock value"""
        return self._lock_dictionary[attribute]
```

```python
"""
file: base layer
"""
    def __setattr__(self, __name: str, __value: Any) -> None:
        if  self.locker.is_locked(__name, is_hard_lock=True):
            print(f"{__name} is locked")
        else:
            return super().__setattr__(__name, __value)
```

**Comment by sofroniewn** [comment]
This is looking very nice!! 

> It is now a hard lock by default as suggested by @tlambert03 and @sofroniewn.

Great! This is now probably out of date (at the top of the code)

```
filename: napari.utils.lock.py
A generic lock class to be used as a locking mechanism. This lock is a non-binding agreement, i.e.,
it is up to the developer to use it and make sure the locks are honoured.
```

> I added lock mode for locking values too. For example, value can be locked to a specific choices or a range.

This is interesting. I like it. It can done differently for each attribute which is cool. I like the initial options you've provided as well.

> there should be some kind of override (either manual or automatic) in case of bugs. Nothing is implemented yet in the code below but I am thinking something that holds the source and if the source is None, we can assume that the plugin crashed and unlock things? Does the plugin widget even goes None when closed?

hmm ok, i need to think about this more. maybe @tlambert03 has an idea

> One thing that I am trying to avoid in this solution is injecting any QT into this class (separation of GUI and Core).

Love this thanks!!

Minor comment: `Lock` above could probably be a [dataclass](https://docs.python.org/3/library/dataclasses.html) or [namedtuple](https://docs.python.org/3/library/collections.html#collections.namedtuple) I think.

One other minor comment. Once we've got this machinery in place we might want to add some convenience methods and GUI buttons to lock/ unlock commonly group attributes (see discussion https://github.com/napari/napari/issues/1570#issuecomment-680759470). That can come much later though

I'd say this looks ready to move to the PR stage. It would be good to add both to the base layer and our eventedmodel (maybe as a `LockableEventedModel`).

Do @jni or @kevinyamauchi want to weigh in on the proposal before then?

**Comment by kevinyamauchi** [comment]
Sorry for the late response! I can have a look later today!

**Comment by kevinyamauchi** [comment]
Hey @MosGeo ! Thank you for the clear and well thought out proposal. I would be interested in seeing this move forward into a PR. I have a few comments/questions below. Thanks again @MosGeo !

- It might make sense to make the `Lock` class a pydantic model so we can validate `Lock.value` against `Lock.value_lock_mode` (e.g., if it's in range mode, make sure the value a list or tuple)
- I am wondering we still want to give the `Lock` an "owner" (i.e., what set the lock) as suggested by @Czaki . I am imaging a situation where a user has multiple plugins installed and they are trying to figure out why they can't change a particular attribute (i.e., who set the lock).
- I think it would be convenient to add a `update` method to `Locker` that allows the user to update multiple locks at once. Perhaps the user passes a dictionary where the keys are the lock names and the values are the new attributes of the lock. What do you think?

**Comment by MosGeo** [comment]
Thanks @sofroniewn and @kevinyamauchi. Sorry for not replying sooner. I didn't get to work on my side projects this weekend, this being one of them. I will incorporate your comments in a draft PR and get back to you. A few notes:

- Ok, I will check out dataclass, namedtuple, and pydantic models. On initial look, I am liking either dataclass (which is basically a normal class) or pydantic models. I have not used pydantic before but I have been reading up on it more and more and it seems powerful.
- While I am still not familiar with the napari eventedmodel yet, i see that it has an setattribute functionality so I am thinking it will be very close to the base layer. The way lock and locker are designed right now is very generic and can be used anywhere (even outside napari 😊 )
- Yes, I think incorporating owner is a very important. I am still not sure what is the best approach. I have to read up on plugin lifecycle before proposing a viable solution.
- Yes, including update and helper methods will be a breeze once the foundation is set :)

**Comment by MosGeo** [comment]
It has been while :)... It seems that there is still interests in locks in napari. I've had some time over the break to play around with this further. Here is the latest update:

[https://github.com/MosGeo/napari/blob/lock_layers/src/napari/utils/lock.py
](https://github.com/MosGeo/napari/blob/lock_layers/src/napari/utils/lock.py
)

I think it covers most of the requested ideas here. Here are the major updates:
- `LockMode` is now a `StrEnum` to allow the user to either use the enum or a direct string in functions.
- `Lock` is now a pydantic model as suggested. The possible values are validated based on the mode automatically.
- Implemented ownership in each lock. By default the ownership is deterimined automatically by determining/saving the name (i.e. string) of the calling package. Currently, by default, even the owner cannot change locked values.
- Removed hard/soft lock. Everything is hard locked now.
- Implemented a context manager `override_lock` to allow overriding the lock temprarly. This means that plugins can open it up, update, and lock it again. Right now, any plugin can open it up (it doesn't respect ownership).

Would love to get some input: @kevinyamauchi @Czaki @tlambert03.

**Comment by TimMonko** [comment]
@MosGeo can you make a PR (even a draft) so that I can play with this. We've been talking (again) about this at the plugin meeting today. Apologies that your comment went unaddressed, I think holidays are a bit to blame, but I did read it at the time and wish I responded sooner! 🙏
